engine/src/cmd/armed.cpp

```diff
@@ -190,11 +190,11 @@ void Armed::ActivateGuns( const weapon_info *sz, bool ms )
 void Armed::Fire( unsigned int weapon_type_bitmask, bool listen_to_owner)
 {
     Unit *unit = static_cast<Unit*>(this);
-    static bool can_fire_in_spec  = GameConfig::GetVariable( "physics", "can_fire_in_spec", false);
-    static bool can_fire_in_cloak = GameConfig::GetVariable( "physics", "can_fire_in_cloak", false);
+    /*static bool can_fire_in_spec  = GameConfig::GetVariable( "physics", "can_fire_in_spec", false);
+    static bool can_fire_in_cloak = GameConfig::GetVariable( "physics", "can_fire_in_cloak", false);*/
     static bool verbose_debug     = GameConfig::GetVariable("data", "verbose_debug", false);
 
-    if ( (unit->cloaking >= 0 && can_fire_in_cloak == false) || (unit->graphicOptions.InWarp && can_fire_in_spec == false) ) {
+    if ( (unit->cloaking >= 0 && weapons_config.can_fire_in_cloak == false) || (unit->graphicOptions.InWarp && weapons_config.can_fire_in_spec == false) ) {
```

Comment:
This can be cached in the object data as it shouldn't change during runtime, however object state might change in relevant ways so only the object's stored version should then get modified
Object can load the information in it's constructor (or a common function if there are multiple constructors or if the object needs to be reset).

Iow it's rare that objects would touch the global object.

Reply:
Just to clarify what you mean - it can be cached within Armed as a field or if it is in only one function, within that function as a static variable?

There was also the question of const-ness. Currently, all settings are loaded from a file and are therefore const. However, in the future, I imagine a GUI to modify these in-game.

Regarding the caching, I'm not sure why we would do such a thing. Wouldn't accessing the config object be  extremely fast?! There's no function at all. You're accessing a variable.

Having said all of this, I actually thought about injecting these during initialization. One of the issues I have encountered when refactoring unit_generic is how interconnected everything in the game is to everything else. The Armed class, for example, doesn't need to know about GameConfig and vice-versa. However, something like this would not be trivial to implement, especially if we factor in changes in configuration mid-game.

At the end of this design review, I'd like to push an update just for Armed, showing the final form of phase 1. It won't be everything I've talked about such as DI and in-game configuration. Just the basic stuff.

Reply:
no, don't use a static variable - use a member variable. This would be faster than accessing the global config all the time as it would be stack local with the object versus having to shunt out to another page to get the global information, invoke another function call, etc.

Reply:
I wouldn't make them constant either; no need for that.

Reply:
I have no idea why we would need caching. This mechanism would be very fast regardless, seems to me.

Reply:
@stephengtuggy we should only access the global objects as few times as possible; caching to a member variable will speed up access b/c it'll be object local memory as opposed to requiring another function call to access.

Reply:
(The (overall) performance impact of caching vs. not caching is going to largely be dependent on the tradeoff between two things: 1) total size of global config structure 2) degree of duplication among object instances + relative increase in object size. 
Considering extreme cases for both is simple: 

1. if the global config is very large and accessed with poor locality and infrequently, then the cost per access will be high for specifically those loads into the global config. If the locality/size/reuse distance is sufficiently bad, then, as Benjamen alluded to, the translation for the page(s) the global config is stored in may have been evicted from the L1 TLB or even the L2 TLB by the time the next field is accessed, which will incur additional overheads
2. if a substantial portion of the total fields of a given object are copies of the global config values and there are many instances of that object that are concurrently active, then the working set may have poor (HW) cache locality compared to a case where only one copy of these globals has to be in the (HW) cache and all of the objects themselves are smaller, which may make the average memory access latency for all memory operations worse.
 
As for the overhead of a call vs. caching and using, that depends on how the global struct is exposed. If it's full-on raw access to a fully public struct, there won't even be a function call - but you've also given away write access. Also, even without there being a function call, there may still be a table indirection, depending on how the global space is being managed. If there's a getter that returns a const ref to the struct, it's not clear that this will actually be compiled to a function call either, since it's really just static checking on the uses of a pointer you passed. If it is, and it's say, in a computationally intense inner loop, then the cached version will likely be much better. Again, however, the total performance impact will be dependent on what fraction of code benefits from caching vs. is penalized for the per-object overheads.

Reply:
Re: local modifications --

While the mechanism would certainly support using copies of global configuration values for locally modifiable fields, I think we should consider whether that's a good paradigm in general to allow it to always be used that way. To that end, perhaps it's worth splitting the sets of global configuration values at the top level into a few bins, and this will make the semantics of each set cleaner:

1. Global constants for the loaded dataset, e.g. physics properties. There are some number of global, dataset specific constants (say, the conversion between VSU damage units and Joules). These are actually constant within the scope of a dataset. 

2. User-specific settings/preferences, e.g. game resolution, default volume. These are game options and, even if they don't (and can't currently) change at run-time, these are logically fungible -- and changes to these values should be global changes, because the properties are global.

3. Globally specified default / useful state values for member variables of objects. These would be used to initialize, reset, adjust member variables... but said member variables are not (at a higher logical semantics level) copy-on-write versions of the global field , but just member values that happen, at some point in time to be set to values from a global field. To give a more concrete example, we might initialize "fuelamount" to a global value for "initial_fuel_amount" - but we wouldn't really want a copy of initial_fuel_amount hanging around, because we already have a member variable that encodes the current state of that local property

Reply:
I suspect we'll see more of case 1 than case 2 per https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/484#discussion_r663310115. We may even end up with mixtures of the two - e.g some code will exhibit case 1 while other code will exhibit case 2. Ultimately we won't know until we do real performance analysis.

That said, if we want to preserve our ability to later do writes on the config, then I suggest we do inline C++ getters instead of publicly exposing the values. For the moment,  it'll compile down to fast getters that will be equivalent of the current code but give us the flexibility to change to change it later easier.

Reply:
I agree - let's do something that uses a clean abstraction and has the features we want first (like: modifiable, but not public), and then worry about the constant factors of the specific performance impact once we can measure it. If there's specific performance critical code that turns out to need to cache values locally, that can still be done on top of the fast getters as the usually-good-enough case (for actually-constants, it's even completely safe) ... and if everything/too much needs caching, then we can always revisit the organization in general.